llvm/lib/Target/X86/X86ISelLowering.cpp

```diff
@@ -56517,17 +56501,40 @@ static SDValue combineFP16_TO_FP(SDNode *N, SelectionDAG &DAG,
 
 static SDValue combineFP_EXTEND(SDNode *N, SelectionDAG &DAG,
                                 const X86Subtarget &Subtarget) {
+  EVT VT = N->getValueType(0);
+  bool IsStrict = N->isStrictFPOpcode();
+  SDValue Src = N->getOperand(IsStrict ? 1 : 0);
+  EVT SrcVT = Src.getValueType();
+
+  SDLoc dl(N);
+  if (SrcVT.getScalarType() == MVT::bf16) {
+    if (!IsStrict && Src.getOpcode() == ISD::FP_ROUND &&
+        Src.getOperand(0).getValueType() == VT)
```

Comment:
Hi, several months late, but I don't think this optimization is correct even the absence of some sort of strictness mode.

That is, `extend(round(x))` isn't meaningfully equal to `x` to a reasonable level of precision - the code I set this off with is explicitly using
```llvm
; This is probably vectorized by the time it hits you
%vRef = load float, ptr %p.iter ; in a loop, etc.
%vTrunc = fptrunc float %vRef to bfloat
%vReExt = fpext bfloat %vTrunc to float
store float %vReExt, ptr %p.iter
```
in order to "lose" extra precision from a f32 computation being used as a reference for a bfloat one.

(The high-level structure goes like this
```
bfloat[N] cTest;
gpu_run(kernel, ..., cTest);
float[N] cTestExt = bfloat_to_float(cTest);
float[N] cRef;
gpu_run(refKernel, ..., cRef);
cRef = float_to_bfloat(bfloat_to_float(cRef));
test_accuracy(cTestExt, cRef, N, ...);
```

I'll also note that this transformation isn't, from what I can tell, present for `half`  

Please revert this conditional specifically or justify it, thanks!

Reply:
There are some discussions in https://github.com/llvm/llvm-project/commit/3cf8535dbf0bf5fafa99ea1f300e2384a7254fba

`bfloat` is different from `half` in two ways:

- `bfloat` has fewer fraction bits, so precision should not be a concern like other types (even `half`) by design;
- `half` is an IEEE type, while `bfloat` is not. We don't necessarily follow it;

Reply:
Discussion noted. However, I as the programmer want to explicitly perform that truncate-extend behavior in one spot in my input (because I'm testing a bfloat function whose result has been `fpext`ed to against a floating-point version that's had its lower bits masked off).

This rewrite has caused per-element result errors around 1e-2 (if I remember right 16.25 vs 16.3125 or the like)

I understand that this intermediate elimination improves performance and is numerically useful a lot of the time, so, given that ... what mechanism would you recommend for forcing this optimization to not fire for a particular pair of round and extend operations?

(I don't want to make things strict at the function level if possible - I want to protect a particular fptrunc / fpext pair from this optimization. Would `canoicalize` do it, or should I stick in s more interesting noop?)

Reply:
> what mechanism would you recommend for forcing this optimization to not fire for a particular pair of round and extend operations?

After https://github.com/llvm/llvm-project/pull/90836 landed, you can use `__arithmetic_fence` to achieve this, e.g., https://godbolt.org/z/vYr1z3h71